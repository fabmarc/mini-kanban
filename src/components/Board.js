import React, { useState } from 'react';
import { DndProvider } from 'react-dnd';
import HTML5Backend from 'react-dnd-html5-backend';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import _ from 'lodash';

import Column from './Column';

const Wrapper = styled.div`
  height: 100vh;
  padding: 16px;
`;

const Header = styled.div`
  height: 58px;
  display: flex;
  flex-direction: column;
  align-items: center;
  font-size: calc(10px + 2vmin);
  font-weight: bold;
  justify-content: center;
  border-bottom: 1px solid #c0c0c0;
`;

const Columns = styled.div`
  height: calc(100% - 58px);
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  padding: 16px;
`;

function Board({ columns, tickets, onChange }) {
  const [allTickets, setAllTickets] = useState(tickets || {});

  const handleTicketChange = (status, tickets) => {
    if (onChange) onChange(status, tickets);
    setAllTickets((prevAllTickets) => {
      prevAllTickets[status] = tickets;
      return prevAllTickets;
    });
  };

  const handleTicketMove = (newStatus, ticket) => {
    setAllTickets((prevAllTickets) => {
      delete prevAllTickets[ticket.status][ticket.id];
      return {
        ...prevAllTickets,
        [newStatus]: {
          ...prevAllTickets[newStatus],
          [ticket.id]: { ...ticket, status: newStatus }
        },
        [ticket.status]: {
          ...prevAllTickets[ticket.status],
        },
      };
    });
  };

  const statuses = Object.keys(columns || {});

  return (
    <DndProvider backend={HTML5Backend}>
      <Wrapper>
        <Header>Mini-Kanban Board</Header>
        <Columns>
          {_.map(statuses, status => (
            <Column
              key={status}
              onDrop={handleTicketMove}
              onChange={handleTicketChange}
              width={`${100 / statuses.length}%`}
              tickets={allTickets[status]}
              header={columns[status]}
              status={status}
            />
          ))}
        </Columns>
      </Wrapper>
    </DndProvider>
  );
}

Board.propTypes = {
  columns: PropTypes.object,
  tickets: PropTypes.objectOf(
    PropTypes.objectOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        description: PropTypes.string.isRequired,
        status: PropTypes.string.isRequired,
      }))),
  onChange: PropTypes.func,
};

export default Board;
