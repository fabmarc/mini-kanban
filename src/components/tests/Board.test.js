import React from 'react';
import { shallow, mount } from 'enzyme';
import HTML5Backend from 'react-dnd-html5-backend';
import renderer from 'react-test-renderer';
import { DndProvider } from 'react-dnd';
import Board from '../Board';

describe('<Board>', () => {
  it('renders without crashing', () => {
    shallow(
      <DndProvider backend={HTML5Backend}>
        <Board />
      </DndProvider>
    );
  });

  it('renders correctly with no columns', () => {
    const tree = renderer
      .create(
        <DndProvider backend={HTML5Backend}>
          <Board />
        </DndProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with columns', () => {
    const tree = renderer
      .create(
        <DndProvider backend={HTML5Backend}>
          <Board columns={{
            todo: 'To-Do',
            ongoing: 'In progress',
            done: 'Done',
          }} />
        </DndProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with tickets', () => {
    const tree = renderer
      .create(
        <DndProvider backend={HTML5Backend}>
          <Board
            columns={{
              todo: 'To-Do',
              ongoing: 'In progress',
              done: 'Done',
            }}
            tickets={{
              todo: {
                1: { id: 1, description: 'Test 1', status: 'todo' },
              },
              ongoing: {
                2: { id: 2, description: 'Test 2', status: 'ongoing' },
                4: { id: 4, description: 'Test 4', status: 'ongoing' },
              },
              done: {
                3: { id: 3, description: 'Test 3', status: 'done' },
              },
            }}
          />
        </DndProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
