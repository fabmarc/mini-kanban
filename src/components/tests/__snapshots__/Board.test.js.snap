// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Board> renders correctly with columns 1`] = `
<div
  className="sc-gzVnrw jpFIFM"
>
  <div
    className="sc-htoDjs kIneFD"
  >
    Mini-Kanban Board
  </div>
  <div
    className="sc-dnqmqq gBTbCw"
  >
    <div
      className="sc-ifAKCX eYfqbg"
      width="33.333333333333336%"
    >
      <div
        className="sc-EHOje kYqLhJ"
      >
        <div
          className="sc-htpNat cXCOUR"
        >
          To-Do
        </div>
        <button
          className="sc-bwzfXH kODISU"
          onClick={[Function]}
        >
          +
        </button>
      </div>
      <div
        className="sc-bZQynM lfIEnT"
      />
    </div>
    <div
      className="sc-ifAKCX eYfqbg"
      width="33.333333333333336%"
    >
      <div
        className="sc-EHOje kYqLhJ"
      >
        <div
          className="sc-htpNat cXCOUR"
        >
          In progress
        </div>
        <button
          className="sc-bwzfXH kODISU"
          onClick={[Function]}
        >
          +
        </button>
      </div>
      <div
        className="sc-bZQynM lfIEnT"
      />
    </div>
    <div
      className="sc-ifAKCX eYfqbg"
      width="33.333333333333336%"
    >
      <div
        className="sc-EHOje kYqLhJ"
      >
        <div
          className="sc-htpNat cXCOUR"
        >
          Done
        </div>
        <button
          className="sc-bwzfXH kODISU"
          onClick={[Function]}
        >
          +
        </button>
      </div>
      <div
        className="sc-bZQynM lfIEnT"
      />
    </div>
  </div>
</div>
`;

exports[`<Board> renders correctly with no columns 1`] = `
<div
  className="sc-gzVnrw jpFIFM"
>
  <div
    className="sc-htoDjs kIneFD"
  >
    Mini-Kanban Board
  </div>
  <div
    className="sc-dnqmqq gBTbCw"
  />
</div>
`;

exports[`<Board> renders correctly with tickets 1`] = `
<div
  className="sc-gzVnrw jpFIFM"
>
  <div
    className="sc-htoDjs kIneFD"
  >
    Mini-Kanban Board
  </div>
  <div
    className="sc-dnqmqq gBTbCw"
  >
    <div
      className="sc-ifAKCX eYfqbg"
      width="33.333333333333336%"
    >
      <div
        className="sc-EHOje kYqLhJ"
      >
        <div
          className="sc-htpNat cXCOUR"
        >
          To-Do
        </div>
        <button
          className="sc-bwzfXH kODISU"
          onClick={[Function]}
        >
          +
        </button>
      </div>
      <div
        className="sc-bZQynM lfIEnT"
      >
        <div
          className="sc-bxivhb gGKUWd"
          onDoubleClick={[Function]}
          opacity={1}
        >
          <div
            className="sc-htpNat cmbrzm"
          >
            Test 1
          </div>
          <button
            className="sc-bwzfXH eUNjes"
            onClick={[Function]}
          >
            -
          </button>
        </div>
      </div>
    </div>
    <div
      className="sc-ifAKCX eYfqbg"
      width="33.333333333333336%"
    >
      <div
        className="sc-EHOje kYqLhJ"
      >
        <div
          className="sc-htpNat cXCOUR"
        >
          In progress
        </div>
        <button
          className="sc-bwzfXH kODISU"
          onClick={[Function]}
        >
          +
        </button>
      </div>
      <div
        className="sc-bZQynM lfIEnT"
      >
        <div
          className="sc-bxivhb gGKUWd"
          onDoubleClick={[Function]}
          opacity={1}
        >
          <div
            className="sc-htpNat cmbrzm"
          >
            Test 2
          </div>
          <button
            className="sc-bwzfXH eUNjes"
            onClick={[Function]}
          >
            -
          </button>
        </div>
        <div
          className="sc-bxivhb gGKUWd"
          onDoubleClick={[Function]}
          opacity={1}
        >
          <div
            className="sc-htpNat cmbrzm"
          >
            Test 4
          </div>
          <button
            className="sc-bwzfXH eUNjes"
            onClick={[Function]}
          >
            -
          </button>
        </div>
      </div>
    </div>
    <div
      className="sc-ifAKCX eYfqbg"
      width="33.333333333333336%"
    >
      <div
        className="sc-EHOje kYqLhJ"
      >
        <div
          className="sc-htpNat cXCOUR"
        >
          Done
        </div>
        <button
          className="sc-bwzfXH kODISU"
          onClick={[Function]}
        >
          +
        </button>
      </div>
      <div
        className="sc-bZQynM lfIEnT"
      >
        <div
          className="sc-bxivhb gGKUWd"
          onDoubleClick={[Function]}
          opacity={1}
        >
          <div
            className="sc-htpNat cmbrzm"
          >
            Test 3
          </div>
          <button
            className="sc-bwzfXH eUNjes"
            onClick={[Function]}
          >
            -
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
